src/state_machine.zig

```diff
@@ -302,186 +291,278 @@ pub const StateMachine = struct {
         }
-        if (!t.flags.condition and !zeroed_32_bytes(t.reserved)) return .reserved_field;
 
-        if (t.amount == 0) return .amount_is_zero;
+        if (t.debit_account_id == 0) return .debit_account_id_must_not_be_zero;
+        if (t.credit_account_id == 0) return .credit_account_id_must_not_be_zero;
+        if (t.credit_account_id == t.debit_account_id) return .accounts_must_be_different;
 
-        if (t.debit_account_id == t.credit_account_id) return .accounts_are_the_same;
+        if (t.pending_id != 0) return .pending_id_must_be_zero;
+        if (t.flags.pending) {
+            // Otherwise, reserved amounts may never be released.
+            if (t.timeout == 0) return .pending_transfer_must_timeout;
+        } else {
+            if (t.timeout != 0) return .timeout_reserved_for_pending_transfer;
+        }
 
-        // The etymology of the DR and CR abbreviations for debit and credit is interesting, either:
-        // 1. derived from the Latin past participles of debitum and creditum, debere and credere,
+        if (t.ledger == 0) return .ledger_must_not_be_zero;
+        if (t.code == 0) return .code_must_not_be_zero;
+        if (t.amount == 0) return .amount_must_not_be_zero;
+
+        // The etymology of the DR and CR abbreviations for debit/credit is interesting, either:
+        // 1. derived from the Latin past participles of debitum/creditum, i.e. debere/credere,
         // 2. standing for debit record and credit record, or
         // 3. relating to debtor and creditor.
         // We use them to distinguish between `cr` (credit account), and `c` (commit).
-        var dr = self.get_account(t.debit_account_id) orelse return .debit_account_not_found;
-        var cr = self.get_account(t.credit_account_id) orelse return .credit_account_not_found;
+        const dr = self.get_account(t.debit_account_id) orelse return .debit_account_not_found;
+        const cr = self.get_account(t.credit_account_id) orelse return .credit_account_not_found;
+        assert(dr.id == t.debit_account_id);
+        assert(cr.id == t.credit_account_id);
         assert(t.timestamp > dr.timestamp);
         assert(t.timestamp > cr.timestamp);
 
-        if (dr.unit != cr.unit) return .accounts_have_different_units;
+        if (dr.ledger != cr.ledger) return .accounts_must_have_the_same_ledger;
+        if (t.ledger != dr.ledger) return .transfer_must_have_the_same_ledger_as_accounts;
+
+        if (t.flags.pending) {
+            if (sum_overflows(t.amount, dr.debits_pending)) return .overflows_debits_pending;
+            if (sum_overflows(t.amount, cr.credits_pending)) return .overflows_credits_pending;
+        }
+        if (sum_overflows(t.amount, dr.debits_posted)) return .overflows_debits_posted;
+        if (sum_overflows(t.amount, cr.credits_posted)) return .overflows_credits_posted;
+        if (sum_overflows(dr.debits_pending, dr.debits_posted)) return .overflows_debits;
```

Comment:
I added this due to a scenario which will not be possible actually. 
```
        mem.zeroInit(Account, .{
            .id = 4,
            .ledger = 1,
            .code = 1,
            .debits_pending = std.math.maxInt(u64) - 1,
            .debits_posted = std.math.maxInt(u64) - 1,
        }),
```

@jorangreef  This will be the case to create the overflow. Do we keep it?

Reply:
Done.